### PR TITLE
Test xleaflet/ipyleaflet widget attribute

### PR DESCRIPTION
**Pull request by VasavanThiru**

### CODE REVIEW

File: include/xleaflet/test_xleaflet.hpp [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+    xeus::xjson state; xeus::buffer_sequence buffers;
+    xleaflet_widget.serialize_state(state, buffers);
+    std::set<std::string> state_set;
+    for (auto& el : state.items()){
+        state_set.insert(el.key());
+    }
+    return state_set;
+}
+
+std::vector<std::string> intersection_widget(std::set<std::string> parsed_set, std::set<std::string> state_set)
```
Comment (martinRenou) [inline include/xleaflet/test_xleaflet.hpp]: You never use this function :)

File: include/xleaflet/test_xleaflet.hpp [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+        return detail::from_json_impl(j);
+    }
+
+    void adl_serializer<py::object>::to_json(json& j, const py::object& obj)
+    {
+        j = detail::to_json_impl(obj);
+    }
+}
+
+std::set<std::string> state_set(auto xleaflet_widget)
```
Comment (martinRenou) [inline include/xleaflet/test_xleaflet.hpp]: You also never use this one :)

File: include/xleaflet/test_xleaflet.hpp [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+    while(it != difference.end())
+    {
+        std::cout << *it << " \n";
+        it++;
+    }
+    std::cout<<std::endl;
+    return difference;
+}
+
+std::set<std::string> set_item(xeus::xjson& item_set)
```
Comment (martinRenou) [inline include/xleaflet/test_xleaflet.hpp]: `std::set<std::string> extract_attribute_names(xeus::xjson& widget_representation)` would be clearer

File: include/xleaflet/test_xleaflet.hpp [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+{
+    std::set<std::string> leaflet_set_item;
+    for(auto& el: item_set.items()){
+        leaflet_set_item.insert(el.key());
+    }
+    return leaflet_set_item;
```
Comment (martinRenou) [inline include/xleaflet/test_xleaflet.hpp]: `leaflet_set_item` -> `attribute_names` ?

File: include/xleaflet/test_xleaflet.hpp [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+}
+
+namespace xlf1
```
Comment (martinRenou) [inline include/xleaflet/test_xleaflet.hpp]: I suppose `xlf` is fine :)

File: include/xleaflet/test_xleaflet.hpp [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+{
+    template <class WT>
+    void test_xleaflet(const std::string& ipyleaflet_widget)
+    {
+        std::cout << ipyleaflet_widget << std::endl;
+        py::object ipyleaflet = py::module::import("ipyleaflet");
+        py::object ipyleaflet_widget_file = ipyleaflet.attr(ipyleaflet_widget.c_str())().attr("get_state")();
+        xeus::xjson ipyleaflet_set = ipyleaflet_widget_file;
+        std::set<std::string> ipyleaflet_set_item = set_item(ipyleaflet_set);
```
Comment (martinRenou) [inline include/xleaflet/test_xleaflet.hpp]: `ipyleaflet_set_item` -> `ipyleaflet_attribute_names`

File: include/xleaflet/test_xleaflet.hpp [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+
+namespace xlf1
+{
+    template <class WT>
+    void test_xleaflet(const std::string& ipyleaflet_widget)
+    {
+        std::cout << ipyleaflet_widget << std::endl;
+        py::object ipyleaflet = py::module::import("ipyleaflet");
+        py::object ipyleaflet_widget_file = ipyleaflet.attr(ipyleaflet_widget.c_str())().attr("get_state")();
+        xeus::xjson ipyleaflet_set = ipyleaflet_widget_file;
```
Comment (martinRenou) [inline include/xleaflet/test_xleaflet.hpp]: `ipyleaflet_widget_file` -> `ipyleaflet_widget_state`

File: include/xleaflet/test_xleaflet.hpp [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+        std::set<std::string> ipyleaflet_set_item = set_item(ipyleaflet_set);
+
+        xeus::xjson state; xeus::buffer_sequence buffers;
+        WT().serialize_state(state, buffers);
+        std::set<std::string> xleaflet_set_item = set_item(state);
+
+        std::vector<std::string> leaflet_difference = difference_widget (ipyleaflet_set_item, xleaflet_set_item);
```
Comment (martinRenou) [inline include/xleaflet/test_xleaflet.hpp]: `EXPECT_EQ(ipyleaflet_attribute_names, xleaflet_attribute_names);`

File: include/xleaflet/test_xleaflet.hpp [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+    while(it != intersection.end())
+    {
+        std::cout << *it << " \n";
+        it++;
+    }
+    std::cout<<std::endl;
+    return intersection;
+}
+
+std::vector<std::string> difference_widget(std::set<std::string> parsed_set, std::set<std::string> state_set)
```
Comment (JohanMabille) [inline include/xleaflet/test_xleaflet.hpp]: Arguments should be const reference

File: include/xleaflet/test_xleaflet.hpp (hunk 3 above)
Comment (JohanMabille) [inline include/xleaflet/test_xleaflet.hpp]: argument should be constant.

File: include/xleaflet/test_xleaflet.hpp [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+        it++;
+    }
+    std::cout<<std::endl;
+    return difference;
+}
+
+std::set<std::string> set_item(xeus::xjson& item_set)
+{
+    std::set<std::string> leaflet_set_item;
+    for(auto& el: item_set.items()){
```
Comment (JohanMabille) [inline include/xleaflet/test_xleaflet.hpp]: `auto&` -> `const auto&`

File: include/xleaflet/test_xleaflet.hpp [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+        return detail::from_json_impl(j);
+    }
+
+    void adl_serializer<py::object>::to_json(json& j, const py::object& obj)
+    {
+        j = detail::to_json_impl(obj);
+    }
+}
+
+std::vector<std::string> difference_widget(const std::set<std::string>& parsed_set, const std::set<std::string>& state_set)
```
Comment (martinRenou) [inline include/xleaflet/test_xleaflet.hpp]: Maybe `compare_attributes` could be a better name, it's a comparison action

File: include/xleaflet/test_xleaflet.hpp [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+    {
+        std::cout << ipyleaflet_widget << std::endl;
+        py::object ipyleaflet = py::module::import("ipyleaflet");
+        py::object ipyleaflet_widget_state = ipyleaflet.attr(ipyleaflet_widget.c_str())().attr("get_state")();
+        xeus::xjson ipyleaflet_set = ipyleaflet_widget_state;
+        std::set<std::string> ipyleaflet_attribute_names = extract_attributes_names(ipyleaflet_set);
+
+        xeus::xjson state; xeus::buffer_sequence buffers;
+        WT().serialize_state(state, buffers);
+        std::set<std::string> xleaflet_set_item = extract_attributes_names(state);
```
Comment (martinRenou) [inline include/xleaflet/test_xleaflet.hpp]: `xleaflet_set_item` -> `xleaflet_attribute_names` for homogeneity with `ipyleaflet_attribute_names`

**Comment by SylvainCorlay** [comment]
Congrats on the merged PR!

**Pull request merged**